include/deal.II/numerics/vector_tools.templates.h

```diff
@@ -800,6 +806,85 @@ namespace VectorTools
     }
 
 
+
+    /**
+     * Generic implementation for the project() function on distributed triangulations.
+     * @p vec_result is expected to not have any ghost entries.
+     */
+    template <int dim, int spacedim, typename VectorType, int components, int fe_degree>
+    void do_project_distributed (const Mapping<dim, spacedim>                              &mapping,
+                                 const DoFHandler<dim, spacedim>                           &dof,
+                                 const ConstraintMatrix                                    &constraints,
+                                 const Quadrature<dim>                                     &quadrature,
+                                 const Function<spacedim, typename VectorType::value_type> &function,
+                                 VectorType                                                &vec_result,
+                                 const bool                                                 enforce_zero_boundary,
+                                 const Quadrature<dim-1>                                   &q_boundary,
+                                 const bool                                                 project_to_boundary_first)
+    {
+      const parallel::Triangulation<dim,spacedim> *parallel_tria =
+        dynamic_cast<const parallel::Triangulation<dim,spacedim>*>(&dof.get_tria());
+      Assert (parallel_tria !=0, ExcNotImplemented());
+      Assert (project_to_boundary_first == false, ExcNotImplemented());
+      Assert (enforce_zero_boundary == false, ExcNotImplemented());
+
+      const MPI_Comm &mpi_communicator = parallel_tria->get_communicator();
+
+      typedef typename VectorType::value_type number;
+      Assert (dof.get_fe().n_components() == function.n_components,
+              ExcDimensionMismatch(dof.get_fe().n_components(),
+                                   function.n_components));
+      Assert (vec_result.size() == dof.n_dofs(),
+              ExcDimensionMismatch (vec_result.size(), dof.n_dofs()));
+      Assert (dof.get_fe().degree == fe_degree,
+              ExcDimensionMismatch(fe_degree, dof.get_fe().degree));
+      Assert (dof.get_fe().n_components() == components,
+              ExcDimensionMismatch(components, dof.get_fe().n_components()));
+
+      // set up mass matrix and right hand side
+      typename MatrixFree<dim,number>::AdditionalData additional_data;
+      additional_data.tasks_parallel_scheme =
+        MatrixFree<dim,number>::AdditionalData::partition_color;
+      additional_data.mpi_communicator = mpi_communicator;
+      additional_data.mapping_update_flags = (update_values | update_JxW_values);
+      MatrixFree<dim, number> matrix_free;
+      matrix_free.reinit (mapping, dof, constraints,
```

Comment:
there is no `MatrixFree.reinit()` for `Mapping<dim,spacedim>` with `dim!=spacedim`. So i wonder if we should only have those functions for dim==spacedim?